server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java

```diff
@@ -288,6 +294,9 @@ public void writeTo(StreamOutput out) throws IOException {
             final DiscoveryNodeRole compatibleRole = role.getCompatibilityRole(out.getVersion());
             out.writeString(compatibleRole.roleName());
             out.writeString(compatibleRole.roleNameAbbreviation());
+            if (out.getVersion().onOrAfter(Version.V_8_0_0)) {
+                out.writeBoolean(compatibleRole.canContainData());
```

Comment:
Once this is backported, do you plan to change this to not use `getCompatibilityRole` and remove that straightaway? (in 7.10.0 I mean)

Reply:
I might be missing something, but I don't think we can remove it straightaway because masters on older versions still need to be told these nodes with the `data_*` roles can hold data, and I think sending the compatible role is the only way? I think we have to wait to remove compatible role until we are no longer communicating with master nodes that do not understand the "can contain data" field.

Reply:
You're not missing something, you're correct. We'll still need it for pre-7.10 versions. I suppose we could remove it for master only since we do specify that only the latest 7.x release can do a rolling upgrade to 8.0, but otherwise we couldn't remove it from 7.x.

Reply:
Okay, I like the idea of removing it from master right away. I'll pick that up in a follow-up.